allennlp/training/optimizers.py

```diff
@@ -151,6 +156,9 @@ def make_parameter_groups(
                     regex,
                 )
 
+        # If the default group is empty, remove it.
+        if len(parameter_groups[-1]["params"]) <= 0:
+            del parameter_groups[-1]
```

Comment:
I also got rid of the change you made to `make_parameter_groups()`. That didn't feel very elegant to me. But then I discovered I had to make this other change instead.

Do you know why this function creates a default parameter group at all? PyTorch optimizers already handle the default case just fine, so I don't get why this is here. But I also didn't want to change code that's been fine for years for no reason, so I left it.

Reply:
Thanks for fixing that! I didn't like the solution I had and am glad you found another way to do it.

> Do you know why this function creates a default parameter group at all? PyTorch optimizers already handle the default case just fine, so I don't get why this is here.

Yes, I'm also noticing in the output of the tests, every time an optimizer is constructed, it will create an extra group (like below), which in our case will be empty as all of our `model_parameters` for this optimizer will be assigned to groups already. I can look into this a bit further to try and see if there is a case where it's not necessary.

```
16:16:34 - INFO - allennlp.training.optimizers - Group 0: ['encoder._module.bias_ih_l0', 'encoder._module.bias_hh_l1', 'encoder._module.bias_hh_l0', 'encoder._module.bias_ih_l1'], {'optimizer_name': 'encoder', 'lr': 0.001}
16:16:34 - INFO - allennlp.training.optimizers - Group 1: ['encoder._module.weight_ih_l0', 'encoder._module.weight_hh_l0', 'encoder._module.weight_ih_l1', 'encoder._module.weight_hh_l1'], {'optimizer_name': 'encoder', 'lr': 0.002}
16:16:34 - INFO - allennlp.training.optimizers - Group 2: [], {}
```

Reply:
I don't think it's necessary to look into it too deeply. The only thing I'm curious about is why we create the default group at all. But arguably I'm in a better position to find out why than you are, since I know the people that wrote that code originally :-)